neuland/shared/TCAViewer.h

```diff
@@ -0,0 +1,251 @@
+        {
+            if (mSafeGuard)
+            {
+                LOG(fatal) << "TCAViewer<out>: Get() function can only be called once in the scope. ";
+            }
+            else
+            {
+                // make sure this function get called every event
+                LogMCEntryNumber();
+                Clear();
+            }
+            return TCAOutput_SafeGuard<DataType>(this);
+        }
+
+        template <typename... Args>
+        void Emplace_back(Args&&... args)
+        {
+            static_assert(mode == out, "TCAViewer<out>::Emplace_back cannot be called in input mode!");
+            if (mTCA == nullptr)
+            {
+                LOG(error) << "TCAViewer<out>: Cannot emplace back a non-existed TClonesArray!";
+            }
+            new ((*mTCA_out)[mTCA_out->GetEntries()]) DataType(std::forward<Args>(args)...);
+        }
+
+        auto SetFirstAddr() -> bool
+        {
+            auto is_success = false;
+            if (mTCA != nullptr && mTCA->Capacity() != 0)
+            {
+                if (dynamic_cast<DataType*>((*mTCA)[0]) == nullptr)
+                {
+                    LOG(fatal) << "TCAViewer<in>: Cannot convert the element in " << mBranchName << " to the type "
+                               << DataType::Class_Name();
+                }
+                else
+                {
+                    mCont = reinterpret_cast<DataType**>(mTCA->GetObjectRef());
+                    is_success = true;
+                }
+            }
+            return is_success;
+        }
+
+        void LogMCEntryNumber()
+        {
+            auto tempEntryNumber = GetMCEntryNumber();
+            if (mMCEntryNumber.has_value())
+            {
+                MCEntryNumDiagnose(tempEntryNumber);
+            }
+            mMCEntryNumber = tempEntryNumber;
+        }
+
+        void MCEntryNumDiagnose(int currentEntryNumber)
+        {
+            if (mMCEntryNumber.value() == currentEntryNumber)
+            {
+                LOG(fatal) << "TCAViewer<out>: Event number diagnositcs fails. Event number is not incremented.";
+            }
+            else if (currentEntryNumber - mMCEntryNumber.value() > 1)
+            {
+                LOG(fatal) << "TCAViewer<out>: Event number diagnositcs fails. Please make sure Get() is called in "
+                              "every event!";
+            }
+        }
+
+        inline auto GetMCEntryNumber() -> int { return eventHeader->GetMCEntryNumber(); }
+
+        std::optional<int> mMCEntryNumber;
+        std::optional<int> mContCapcitySize;
+        std::string mBranchName;
+        DataType** mCont = nullptr;
+        TClonesArray* mTCA = nullptr;
+        bool mSafeGuard = false;
+        std::unique_ptr<TClonesArray> mTCA_out; // owning
+        FairEventHeader* eventHeader = nullptr; // non-owning
+    };
+
+    template <typename Datatype>
+    class TCAOutput_SafeGuard
```

Comment:
I think I get what this is for, but you do need a small comment on why it is needed.

Reply:
ok, yeah, I should do that.